src/env.h

```diff
@@ -770,6 +778,10 @@ class AsyncHooks : public MemoryRetainer {
 
   // Non-empty during deserialization
   const SerializeInfo* info_ = nullptr;
+
+  std::vector<v8::Global<v8::Context>> contexts_;
```

Comment:
Ideally we should implement snapshot support for these, but since they aren't used by default in the bootstrap, I think we could just CHECK that these are empty in `AsyncHooks::Serialize` to indicate that these are not yet supported (I am not sure if the support of promise hooks in snapshots is implemented on the V8 side either though).

Reply:
@joyeecheung There will be one (weak ref'd) context in the vector at snapshot time (the main context). Do you mean the promise hook functions?

e.g.

```diff
diff --git a/src/env.cc b/src/env.cc
index 042b7243ac..f4d789710e 100644
--- a/src/env.cc
+++ b/src/env.cc
@@ -1156,6 +1156,11 @@ AsyncHooks::SerializeInfo AsyncHooks::Serialize(Local<Context> context,
         context,
         native_execution_async_resources_[i].Get(context->GetIsolate()));
   }
+  CHECK_EQ(contexts_.size(), 1);
+  CHECK(js_promise_hooks_[0].IsEmpty());
+  CHECK(js_promise_hooks_[1].IsEmpty());
+  CHECK(js_promise_hooks_[2].IsEmpty());
+  CHECK(js_promise_hooks_[3].IsEmpty());

   return info;
 }
 ```

Reply:
Ah, yes, that looks about right (maybe for additional safety, you could check that one context is indeed the main context)

Reply:
Alright, done in latest fixup commit 👍